diff --git a/drivers/net/ethernet/rockchip/gmac/stmmac_main.c b/drivers/net/ethernet/rockchip/gmac/stmmac_main.c
index 5d2df5b4ffb..ba5d0f83886 100755
--- a/drivers/net/ethernet/rockchip/gmac/stmmac_main.c
+++ b/drivers/net/ethernet/rockchip/gmac/stmmac_main.c
@@ -925,8 +925,8 @@ static int stmmac_init_phy(struct net_device *dev)
        }

        /* Stop Advertising 1000BASE Capability if interface is not GMII */
-       if ((interface == PHY_INTERFACE_MODE_MII) ||
-           (interface == PHY_INTERFACE_MODE_RMII))
+       //if ((interface == PHY_INTERFACE_MODE_MII) ||
+       //    (interface == PHY_INTERFACE_MODE_RMII))
                phydev->advertising &= ~(SUPPORTED_1000baseT_Half |
                                         SUPPORTED_1000baseT_Full);



强的这修改，再读和验证一下
sdk_project@aaaaa:~/5.1_sdk/RK3288_ANDROID5.1-SDK_V1.00_20150515/kernel/drivers/net/phy$ git diff .
diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index 3657b4a..514b2d2 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -871,7 +871,7 @@ int genphy_read_status(struct phy_device *phydev)
                phydev->pause = phydev->asym_pause = 0;

                if (lpagb & (LPA_1000FULL | LPA_1000HALF)) {
-                       phydev->speed = SPEED_1000;
+                       phydev->speed = SPEED_100;

                        if (lpagb & LPA_1000FULL)
                                phydev->duplex = DUPLEX_FULL;
@@ -907,7 +907,7 @@ int genphy_read_status(struct phy_device *phydev)

                phydev->pause = phydev->asym_pause = 0;
        }
-
+
        return 0;
 }
 EXPORT_SYMBOL(genphy_read_status);
@@ -952,7 +952,7 @@ static int genphy_config_init(struct phy_device *phydev)
                if (val & ESTATUS_1000_THALF)
                        features |= SUPPORTED_1000baseT_Half;
        }
-
+    features &= ~ (SUPPORTED_1000baseT_Full | SUPPORTED_1000baseT_Half);
        phydev->supported = features;
        phydev->advertising = features;
        
        
        
        
使用10M
diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index 8179727..1d54120 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -161,14 +161,14 @@ struct phy_device *phy_device_create(struct mii_bus *bus, int addr, int phy_id,
 
 	dev->dev.release = phy_device_release;
 
-	dev->speed = 0;
-	dev->duplex = -1;
+	dev->speed = SPEED_10;
+	dev->duplex = DUPLEX_FULL;
 	dev->pause = 0;
 	dev->asym_pause = 0;
 	dev->link = 1;
 	dev->interface = PHY_INTERFACE_MODE_GMII;
 
-	dev->autoneg = AUTONEG_ENABLE;
+	dev->autoneg = AUTONEG_DISABLE;
 
 	dev->is_c45 = is_c45;
 	dev->addr = addr;
@@ -1108,6 +1108,9 @@ int genphy_read_status(struct phy_device *phydev)
 		else
 			phydev->speed = SPEED_10;
 
+		
+		phydev->duplex = DUPLEX_FULL;
+		phydev->speed = SPEED_10;
 		phydev->pause = 0;
 		phydev->asym_pause = 0;
 	}

