diff --git a/arch/arm64/boot/dts/rockchip/rk1808-evb.dtsi b/arch/arm64/boot/dts/rockchip/rk1808-evb.dtsi
index ee31e60957ce..ee699d8f2b24 100644
--- a/arch/arm64/boot/dts/rockchip/rk1808-evb.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk1808-evb.dtsi
@@ -251,16 +251,28 @@
 
 &gmac {
 	phy-supply = <&vcc_phy>;
-	phy-mode = "rgmii";
+	phy-mode = "rmii";
 	clock_in_out = "input";
 	snps,reset-gpio = <&gpio0 10 GPIO_ACTIVE_LOW>;
 	snps,reset-active-low;
 	/* Reset time is 20ms, 100ms for rtl8211f */
 	snps,reset-delays-us = <0 20000 100000>;
-	assigned-clocks = <&cru SCLK_GMAC>;
-	assigned-clock-parents = <&gmac_clkin>;
-	tx_delay = <0x50>;
-	rx_delay = <0x3a>;
+	//assigned-clocks = <&cru SCLK_GMAC>;
+	//assigned-clock-parents = <&gmac_clkin>;
+	clocks = <&cru SCLK_GMAC>, <&cru SCLK_GMAC_RX_TX>,
+		 <&cru SCLK_GMAC_RX_TX>, <&cru SCLK_GMAC_REF>,
+		 <&cru SCLK_GMAC_REFOUT>, <&cru ACLK_GMAC>,
+		 <&cru PCLK_GMAC>, <&cru SCLK_GMAC_RMII_SPEED>;
+	clock-names = "stmmaceth", "mac_clk_rx",
+		 "mac_clk_tx", "clk_mac_ref",
+		 "clk_mac_refout", "aclk_mac",
+		 "pclk_mac", "clk_mac_speed";
+	assigned-clocks = <&cru SCLK_GMAC_RX_TX>, <&cru SCLK_GMAC>;
+	assigned-clock-parents = <&cru SCLK_GMAC_RMII_SPEED>, <&gmac_clkin>;
+};
+
+&gmac_clkin {
+	clock-frequency = <50000000>;
 };
 
 &i2c0 {
