diff --git a/arch/arm64/boot/dts/rockchip/rk3399-sapphire.dtsi b/arch/arm64/boot/dts/rockchip/rk3399-sapphire.dtsi
index 5728818..a2a75a6 100644
--- a/arch/arm64/boot/dts/rockchip/rk3399-sapphire.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3399-sapphire.dtsi
@@ -87,12 +87,12 @@
default-brightness-level = <200>;
};

- clkin_gmac: external-gmac-clock {
+/* clkin_gmac: external-gmac-clock {
compatible = "fixed-clock";
clock-frequency = <125000000>;
clock-output-names = "clkin_gmac";
#clock-cells = <0>;
- };
+ };*/

dw_hdmi_audio: dw-hdmi-audio {
status = "disabled";
@@ -236,12 +236,13 @@
&gmac {
phy-supply = <&vcc_phy>;
phy-mode = "rgmii";
- clock_in_out = "input";
+ clock_in_out = "output";
snps,reset-gpio = <&gpio3 15 GPIO_ACTIVE_LOW>;
snps,reset-active-low;
snps,reset-delays-us = <0 10000 50000>;
- assigned-clocks = <&cru SCLK_RMII_SRC>;
- assigned-clock-parents = <&clkin_gmac>;
+ assigned-clocks = <&cru SCLK_MAC>, <&cru SCLK_RMII_SRC>;//重新设置一下这个两个clk的provider
+ assigned-clock-rates = <125000000>;
+ assigned-clock-parents = <&cru PLL_NPLL>, <&cru SCLK_MAC>;
pinctrl-names = "default";
pinctrl-0 = <&rgmii_pins>;
tx_delay = <0x28>;
diff --git a/arch/arm64/boot/dts/rockchip/rk3399-vop-clk-set.dtsi b/arch/arm64/boot/dts/rockchip/rk3399-vop-clk-set.dtsi
index f13985c..489592b 100644
--- a/arch/arm64/boot/dts/rockchip/rk3399-vop-clk-set.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3399-vop-clk-set.dtsi
@@ -148,7 +148,7 @@
<50000000>, <100000000>,
<75000000>, <75000000>,
<816000000>, <816000000>,
- <600000000>, <200000000>,
+ <1000000000>, <200000000>,
<800000000>, <150000000>,
<75000000>, <37500000>,
<300000000>, <100000000>,
diff --git a/drivers/clk/clk-conf.c b/drivers/clk/clk-conf.c
index 21aee0c..13f8baa 100644
--- a/drivers/clk/clk-conf.c
+++ b/drivers/clk/clk-conf.c
@@ -62,6 +62,8 @@ static int __set_clk_parents(struct device_node *node, bool clk_supplier)
goto err;
}

+ pr_err("wxt_clk: clk %s to pclk:%s\n",
+ __clk_get_name(clk), __clk_get_name(pclk));
rc = clk_set_parent(clk, pclk);
if (rc < 0)
pr_err("clk: failed to reparent %s to %s: %d\n",
